lib/logstash/inputs/snmptrap.rb

```diff
@@ -47,26 +47,44 @@ def register
     end
   end # def register
 
-  public
   def run(output_queue)
     begin
       # snmp trap server
       snmptrap_listener(output_queue)
     rescue => e
       @logger.warn("SNMP Trap listener died", :exception => e, :backtrace => e.backtrace)
-      sleep(5)
-      retry
+      stoppable_call(20) { sleep(0.25) }
```

Comment:
The pipeline thread is calling stop on the input instance, so AFAIK the current thread is not the input thread (which might be sleeping).  Is it true to say that in the pipeline code the input_thread and the input instance do not know about each other?  If so then, in the pipeline thread, when calling the stop method, the instance would need to hold a reference to the input thread to wake itself up OR the pipeline should wakeup the input thread before calling stop on the input instance.

See [this gist](https://gist.github.com/guyboertje/136295a585fd97fbdc81)